packages/std/src/math/fraction.rs

```diff
@@ -12,3 +17,86 @@ pub trait Fraction<T>: Sized {
     /// If `p` is zero, None is returned.
     fn inv(&self) -> Option<Self>;
 }
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct Fraction<T>(T, T);
+
+impl<T> Fraction<T> {
+    pub fn new(numerator: T, denominator: T) -> Self {
+        Self(numerator, denominator)
+    }
+}
+
+impl<T: Clone> Fractional<T> for Fraction<T> {
+    fn numerator(&self) -> T {
+        self.0.clone()
+    }
+
+    fn denominator(&self) -> T {
+        self.1.clone()
+    }
+
+    fn inv(&self) -> Option<Self> {
+        unimplemented!()
+    }
+}
+
+pub trait FractionMath {
+    fn checked_mul_floored<F: Fractional<T>, T: Into<Uint512>>(
+        self,
+        rhs: F,
+    ) -> Result<Self, CheckedMultiplyFractionalError>
+    where
+        Self: Sized;
+
+    fn mul_floored<F: Fractional<T>, T: Into<Uint512>>(self, rhs: F) -> Self;
+
+    fn checked_mul_ceil<F: Fractional<T> + Clone, T: Into<Uint512>>(
+        self,
+        rhs: F,
+    ) -> Result<Self, CheckedMultiplyFractionalError>
+    where
+        Self: Sized;
+
+    fn mul_ceil<F: Fractional<T> + Clone, T: Into<Uint512>>(self, rhs: F) -> Self;
+}
+
+#[macro_export]
+macro_rules! fraction_math {
+    ($name:ident) => {
+        impl FractionMath for $name {
+            fn checked_mul_floored<F: Fractional<T>, T: Into<Uint512>>(
+                self,
+                rhs: F,
+            ) -> Result<Self, CheckedMultiplyFractionalError> {
+                let res = Uint512::from(self)
+                    .checked_mul(rhs.numerator().into())?
```

Comment:
Here we should use full_mul, which is a multiplication that cannot fail. Uint64*Uint64 -> Uint128, Uint128*Uint128 -> Uint256 and so on.

The reason is that the multiplication might exceed the range but after the division, the result is back in range. E.g. if you do `Uint128::MAX * 99/100` should succeed.

You can see that in the `multiply_ratio` implementations.